doc/design/cpp_data_feeding.md

```diff
@@ -69,10 +113,59 @@ To solve this problem, we introduce `ReaderHolder` as a wrapper. It acts as an e
 
 To create and invoke readers, some new ops are introduced:
 
-### `CreateReaderOp`
+### CreateReaderOp
 
 Each reader has its creation op. File readers' creation ops have no input and yield the created file reader as its output. Decorated readers' creation ops take the underlying readers as inputs and then yield new decorated readers.
 
-### `ReadOp`
+However, direct usage of file readers' creation ops is not recommended because a file reader can only read one file via a single thread. Using `OpenFilesOp` is a better choice.
+
+### OpenFilesOp
+
+The `OpenFilesOp` is the creation op of `MultipleReader`. It takes no input but requires a list of file names as one of its attributes. The newly created `MultipleReader` then creates its own prefetching readers according to given file names.
+
+To make sure that created prefetching readers match file formats, we need a name prefix rule to append file format tags to file names, as well as a file reader registry mechanism to map file format tags to their corresponding file readers' constructors.
+
+### HasNextOp
+
+`HasNextOp` is used to check whether the next data batch exists via the reader's `HasNext()` interface.
+
+### ResetOp
+
+`ResetOp` is used to reset a reader via its `ReInit()` interface.
+
+### ReadOp
 
 A reader is only a Variable. It cannot trigger the reading process by itself. So we add the `ReadOp` to execute it. A `ReadOp` takes a reader Variable as its input. Each time it runs, it invokes the reader‘s `ReadNext()` function and gets a new batch of data(or only one instance of data, if we use file reader directly). The output data of a reader are in the form of `std::vector<LoDTenosr>`, so the `ReadOp` also needs to split the vector and move LoDTensors to their respective output Variables.
+
+## Program with Readers
+
+A `Program` holds readers as its persistable variables. These variables are created by `CreateReaderOp` or `OpenFilesOp`. Obviously, these ops shall run only once. So they shall be settled in the `startup_program`. `HasNextOp`, `ResetOp` and `ReadOp` are required by training loop, so they shall be in the `main_program`.
+
+The ops of a `startup_program` with readers would be like this:
+
+```
+multiple_reader = open_files_op(...)
+batch_reader = create_batch_reader_op(multiple_reader)
+double_buffer_reader = create_double_buffer_op(batch_reader)
+... (other initializers)
+```
+
+The forwarding ops of the corresponding `main_program` would be like this:
+
+```
+while_op {
+    has_next = has_next_op(double_buffer_reader)
```

Comment:
@helinwang You are right! We have discussed this issue with @JiayiFeng  and @panyx0718 before. We will have a `MultiEpochReader` for users. Users may not need to use `reset` and `has_next` operators. We will not implement `reset` and `has_next` operators at first.

The concern about `reset` and `has_next` operators is that users might need to print some log and save models when an epoch is ended. It might be useful for users to know an epoch is ended by `has_next_op` and manually reset reader by `reset_op`. However, we will not implement these operators until there are actual cases.

Reply:
Great idea. The `multi_epoch_reader` is also in our design. It is definitely a convenient way for users to config their readers.

But `reset_op` and `has_next_op` still need to be provided. For some users would like to check whether a pass is completed by themselves and then do some customized operations. Just like my reply to your previous comment.

Reply:
Thanks for the reply @reyoung @JiayiFeng !

> But reset_op and has_next_op still need to be provided. For some users would like to check whether a pass is completed by themselves and then do some customized operations. Just like my reply to your previous comment.

That's a good point, but I think we will probably rely on visual DL for visualization. If the user need to do some customize operation, he need to write an OP to do it, which adds the complexity. I my opinion `reset_op`, `has_next_op` and `if op` is a lot of code for the user to learn, maybe we should not support it?